api federation types

deads2k opened the pull request:
First commit adds types that can back the kubernetes-discovery server with an `kubectl` compatible way of adding federated servers.  Second commit is just generated code.

After we have types, I'd like to start splitting `kubernetes-discovery` into a "legacy" mode which will support what we have today and a "normal" mode which will provide an API federation server like this: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/federated-api-servers.md that includes both discovery and proxy in a single server.  Something like this: https://github.com/openshift/kube-aggregator .

@kubernetes/sig-api-machinery @nikhiljindal

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37561)
<!-- Reviewable:end -->

sttts commented on apifederator/pkg/apis/apifederation/helpers.go:
```diff
@@ -0,0 +1,59 @@
+	"strings"
+)
+
+func SortByGroup(servers []*APIServer) [][]*APIServer {
```
nit: SortedByGroup.

It's not in place.

sttts commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// APIServerSpec contains information for locating and communicating with a server.
+// Only https is supported, though you are able to disable host verification.
```
not only host verification, also the whole cert check.

sttts commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+	// InsecureSkipTLSVerify disables TLS certificate verification when communicating with this server.
+	// This is strongly discouraged.  You should use the CABundle instead.
+	InsecureSkipTLSVerify bool
+	// CABundle is a PEM encoded CA bundle which be used to validate an API server's serving certificate.
```
which will be used

sttts commented on apifederator/pkg/apis/apifederation/validation/validation.go:
```diff
@@ -0,0 +1,90 @@
+		},
+		field.NewPath("metadata"))
+
+	// in this case we all empty group
```
all?

sttts commented on apifederator/pkg/apis/apifederation/validation/validation.go:
```diff
@@ -0,0 +1,90 @@
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "insecureSkipTLSVerify"), apiServer.Spec.InsecureSkipTLSVerify, "may not be true if caBundle is present"))
+	}
+	if !apiServer.Spec.InsecureSkipTLSVerify && len(apiServer.Spec.CABundle) == 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "caBundle"), apiServer.Spec.CABundle, "must be present if insecureSkipTLSVerify is missing"))
```
Which means you don't want to allow root cert bundle verification of non-self-signed certs?

deads2k replied:
> Which means you don't want to allow root cert bundle verification of non-self-signed certs?

fixed

lavalamp commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+	// Group is the API group name this server hosts
+	Group string
+	// Version is the API version this server hosts
+	Version string
```
Shouldn't this be a list?

deads2k replied:
> Shouldn't this be a list?

I don't think we want to make this a list.  Thinking about how I want to roll out a second version, I don't want to risk my `v1` server to run a `v2` server.  That means I'm looking at a side-by-side during the transition period, which means GroupVersion granularity instead group granularity.

liggitt replied:
does "side-by-side" mean more than one server could serve the same group/version?

do we want to have to repeat the whole object with all the server-specific bits if a single server is serving multiple groups and/or versions?

liggitt replied:
>> does "side-by-side" mean more than one server could serve the same group/version?
>
> No. One service for v1, another service for v2.

how do you plan to enforce that in case of conflicts? oldest wins?

deads2k replied:
> how do you plan to enforce that in case of conflicts? oldest wins?

Highest priority.

lavalamp replied:
Sure, you *could* do it that way, but I don't think we should *force* people to do it that way.

deads2k replied:
Since the name of the object is forced to be `version.group` (see validation rules), you can't end up with two trying to run the same API version.

deads2k replied:
> Sure, you could do it that way, but I don't think we should force people to do it that way.

I think that the consistent rule of "one groupversion, one resource" is very easy to describe, enforce, prioritize, and avoid conflicts for.  When you start combining them, you open the API enough to be self-contradictory.  I don't think that avoiding granular objects is worth it.  I don't think its likely to be very onerous and prescriptiveness can help tooling built around it.

lavalamp replied:
As long as it's explicitly acceptable to point multiple of these records at the same backing service, I guess I'm OK with this. It does mean there'll be more configuration work involved.

(I was actually expecting to split only by group. I guess it does seem valuable to host different versions separately in some situations, but is it really necessary?)

deads2k replied:
> (I was actually expecting to split only by group. I guess it does seem valuable to host different versions separately in some situations, but is it really necessary?)

This came out of my first encounter with the enemy as I was dealing with an openshift project API that had one version hosted in openshift and another hosted in a third server that works against bare kube.  It got me thinking about what I wanted to do in an upgrade scenario and I realized I wanted to avoid risking my v1 server.

lavalamp commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+
+// APIServerSpec contains information for locating and communicating with a server.
+// Only https is supported, though you are able to disable certificate verification.
+type APIServerSpec struct {
```
This actually (potentially) represents a APIServer replica pool in HA setups. Name is misleading.

lavalamp replied:
So maybe APIService{Spec, Status} is a better name.

deads2k replied:
> So maybe APIService{Spec, Status} is a better name.

doing

lavalamp commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+// Only https is supported, though you are able to disable certificate verification.
+type APIServerSpec struct {
+	// InternalHost is the host:port for locating the server inside the pod/service network.
+	InternalHost string
```
If it's a Host:Port pair, call it an address.

But why present it like this? Why not make this take a service name?

deads2k replied:
> But why present it like this? Why not make this take a service name?

I wasn't prepared to say that it *must* point to a particular service.  An address allows an external reference more easily and some of the first that I wanted to federate weren't in-cluster to start.

lavalamp replied:
That's a fair point, but you can represent such things as a service by adding the endpoint(s) yourself. I think referencing a service will let us understand the topology in a way that a single address will not.

deads2k replied:
> That's a fair point, but you can represent such things as a service by adding the endpoint(s) yourself. I think referencing a service will let us understand the topology in a way that a single address will not.

So you'd like a service reference plus a port?

lavalamp replied:
If you think it's too limiting to ask everyone to use 443, then I guess we need a port.

deads2k replied:
> If you think it's too limiting to ask everyone to use 443, then I guess we need a port.

As a service, I guess there's not a good reason to allow anything else.

deads2k replied:
> But why present it like this? Why not make this take a service name?

doing

lavalamp commented:
So the goal here is to add an api interface to the discovery merger?

Can you move the apifederator directory under cmd/kubernetes-discovery? In fact I think it should just be `cmd/kubernetes-discovery/apis/...`. I'm also in favor of moving the binary into the staging directory.

deads2k commented:
> So the goal here is to add an api interface to the discovery merger?

I want that *and* I want to use the same API as the basis for extending the `kubernetes-discovery` to include a proxy.  I've written up some idea here: https://github.com/kubernetes/kubernetes/pull/37650 .

I'm in favor of the move as well, but I want to wait until the genericapiserver gets there.  I have several pulls in the queue to make genericapiserver easy enough to re-use to drive the `kubernetes-discovery` from it.

lavalamp commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+
+	// InsecureSkipTLSVerify disables TLS certificate verification when communicating with this server.
+	// This is strongly discouraged.  You should use the CABundle instead.
+	InsecureSkipTLSVerify bool
```
Why would that be configurable per-apiserver?

deads2k replied:
> Why would that be configurable per-apiserver?

You could have different custom CAs per-apiserver.  It's less likely, but still possible that you have a few black sheep (unverifiable) amongst the rest.

lavalamp replied:
Anyway, I would say that the summarizer "owns" the discovery paths.

deads2k replied:
> Anyway, I would say that the summarizer "owns" the discovery paths.

You face a similar decision with the proxier when it sets up TLS client verification per-API server.

lavalamp replied:
Let me put it another way: will there ever be a situation in which a client will trust only a subset of the CAs that sign the apiservers's certs? I maintain that that's a configuration error, and I think it will be vastly simpler if there's a single cert that the discovery merger presents.

deads2k replied:
> Let me put it another way: will there ever be a situation in which a client will trust only a subset of the CAs that sign the apiservers's certs? I maintain that that's a configuration error, and I think it will be vastly simpler if there's a single cert that the discovery merger presents.

The discovery merger presents one serving certificate and it isn't this one.  This CA allows the discovery merger and proxier to trust the serving certificates of the backing API servers it is communicating with.

liggitt replied:
> CAs won't be 1:1 with apiservers

in the simplest case (each one self-signs a cert), they are

lavalamp replied:
The only use case I can see for that is a short dev/test cycle for apiservers themselves. I think in any real cluster, the sysadmin is going to want everything signed by the same CA? Is there really a good use case for multiple CAs?

I mean, aren't we adding a certificates resource, so it should be easy to get a cert signed by the main apiserver?

lavalamp commented on apifederator/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
+	CABundle []byte
+
+	// Priority controls the ordering of this API group in the overall discovery document that gets served.
```
Allowing sysadmins to adjust it means kubectl will function differently on different clusters. Big user surprise. I really think we should come up with a better system.

deads2k replied:
> Allowing sysadmins to adjust it means kubectl will function differently on different clusters. Big user surprise. I really think we should come up with a better system.

Pushing it to the client means that a cluster-admin can't choose which groups he wants to have priority in his cluster.  Imagining two groups with `Roles`.  I don't think that a client should choose which one of the `Roles` takes priority by default.  As an explicit choice certainly (we support this today), but by default I think that's unexpected.  I think its a better idea for a  cluster-admin wants to control resource preference in his cluster and a client can choose to be explicit (best practice recommended today).

lavalamp replied:
I didn't say we should push it to the client, I agree, that makes it worse. I do think the resolution order should be a fixed property of the GVKs involved, not a property of the configuration in the discovery merger.

lavalamp replied:
I'm still thinking about this. I agree there's a problem, but I don't like any solution so far. The solution in this PR is high-cognitive overhead for cluster administrators. I would prefer a solution that "just works" but without hard coding something or a central registry somewhere, I haven't thought of anything yet.

lavalamp commented:
I'm pretty strongly opposed to starting a new top-level directory with api types for a binary that already has a directory.

deads2k commented:
> I'm pretty strongly opposed to starting a new top-level directory with api types for a binary that already has a directory.

Ok, I don't mind moving it.

lavalamp commented:
OK, so one more thing to think about: one of the problems with TPR is that they're global, not namespaced. I think people will want to lock some apiservers into some namespaces. For dev/test cycles if no other reason.

liggitt commented:
> lock some apiservers into some namespaces

I'm not sure what you mean by that. A federated API that routes type A to one backend server inside namespace foo but a different backend server inside namespace bar?

lavalamp commented:
Yeah, I think step one is to figure out what people need.

deads2k commented:
> OK, so one more thing to think about: one of the problems with TPR is that they're global, not namespaced. I think people will want to lock some apiservers into some namespaces. For dev/test cycles if no other reason.

I don't think that a summarizer for the cluster as a whole should accept input from someone other than a cluster-admin.  That means that I don't think the scenario of a namespace (project) admin adding a TPR which is reflected in the cluster-wide discovery summarizer is a reasonable one.

If they want to run their own API server hosting a new resource along with an etcd to back it, that seems fine.  I don't think we should expose the core etcd to the whims of a namespace admin.

I would also claim that any "extra" summarizing API server which could be layered on in a namespace would operate differently than this, since it could not be trusted with front proxy credentials.

deads2k commented:
I just completed the rename, so all the commends went away.  I forced the rename since it's boring, but all other changes are in another commit.  Highlights:
 1. moved under existing package
 1. changed name to APIService
 1. point at a service
 1. left priority and CA

deads2k committed:
refactor kubernetes-discovery to use cobra and prep for genericapiserver

deads2k committed:
add basic wiring for kubernetes-discovery to become a kube-like api server

deads2k commented:
@lavalamp I'd really like move this forward today.  Even when merged, it will be an alpha API and few weeks out from being "normal" for any integration.  Can we let it play it out from here?

smarterclayton commented:
So concern about naming overlapping with cluster federation.  I'd prefer not to have to explain myself every time I mentioned the group ("api, not cluster").  One suggestion might be "apiregistration" (in that you are creating registration of APIs, which someone can use).

smarterclayton commented:
Would clusters also have to federate APIs?  If so, will there be the same API, or different APIs?  Will the cluster federator depend on each cluster to federate APIs?

deads2k commented:
I'm fine with apiregistration.  Cluster federation is different since the behavior is different.

Anything other than name?

smarterclayton commented on cmd/kubernetes-discovery/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+
+	// Priority controls the ordering of this API group in the overall discovery document that gets served.
+	// Client tools like `kubectl` use this ordering to derive preference, so this ordering mechanism is important.
+	// Secondary ordering is performed based on name.
```
Detail what the ordering should be based on (ordered by lowest priority first, then by lexical order of the name).

lavalamp replied:
If we have to have this in...

I suggest calling it Tier. I suggest giving examples, like *.k8s.io should be tier 100, PaaSes (OpenShift, Deis) are recommended to be 200, and random addons tier 300.

I suggest limiting it to something like 0-400 or 0-500. 2^64 priorities are NOT necessary.

Doc comment should state that lower number = higher priority.

Secondary is *alphabetical* by name. Arguably that's not the right way to sort domain names, but at least it's obvious what the system will do.

smarterclayton commented on cmd/kubernetes-discovery/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+	// This is strongly discouraged.  You should use the CABundle instead.
+	InsecureSkipTLSVerify bool
+	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
+	CABundle []byte
```
Is there an analogue for this in our admission control plugin config (for image webhook)?

deads2k replied:
That's a file ref because it's configured via a file.  I don't think the mechanism is appropriate here

lavalamp replied:
forgive me if my cert knowledge is wrong, but...

The purpose of this is to tell the merger that it can trust the backend, right? I guess I would have expected this to just be a single public key that merger can trust, and the apiserver service provides the bundle?

Along similar lines, do we need to give merger a client cert to use, or some other way of expressing who it is to the apiserver?

deads2k replied:
> The purpose of this is to tell the merger that it can trust the backend, right? I guess I would have expected this to just be a single public key that merger can trust, and the apiserver service provides the bundle?

The single bundle to initialize the process case *may* be the most common production scenario, but the first experience people are likely to have is the self-signed or individually signed cases which lend themselves to this.



> Along similar lines, do we need to give merger a client cert to use, or some other way of expressing who it is to the apiserver?

Yes.  A request header proxy client certificate trusted by every backend server.  There will be only one of these, since the trust is distributed via a known good client cert.  Once I get enough pieces, I'll wire up `local-up-cluster.sh` to give a good example and try to start wiring it for testing.

smarterclayton commented on cmd/kubernetes-discovery/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+	InternalHost string
+	// Group is the API group name this server hosts
+	Group string
+	// Version is the API version this server hosts
```
Give an example ("v1")

smarterclayton commented on cmd/kubernetes-discovery/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+type APIServerSpec struct {
+	// InternalHost is the host:port for locating the server inside the pod/service network.
+	InternalHost string
+	// Group is the API group name this server hosts
```
Is this allowed to be empty?

deads2k replied:
Only for version "v1"

smarterclayton commented on cmd/kubernetes-discovery/pkg/apis/apifederation/types.go:
```diff
@@ -0,0 +1,71 @@
+// Only https is supported, though you are able to disable certificate verification.
+type APIServerSpec struct {
+	// InternalHost is the host:port for locating the server inside the pod/service network.
+	InternalHost string
```
Seems arbitrary (here) to say that this must be in the pod / service network.  If there's more of a reason, explain here.  If there isn't a strong reason yet, turn it into a recommendation or a suggestion.

deads2k replied:
Request from @lavalamp, but I don't feel strongly.

deads2k replied:
It's later restricted to service ref

smarterclayton replied:
Like we plan on restricting it later?

deads2k replied:
No, two commits down.

lavalamp commented:
I agree, please change the name.

smarterclayton commented:
"Tier" feels weird and unusual (to a casual user).  Weight might be a good
generic word.  But I'm ok with priority.

On Mon, Dec 5, 2016 at 6:54 PM, Daniel Smith <notifications@github.com>
wrote:

> *@lavalamp* commented on this pull request.
> ------------------------------
>
> In cmd/kubernetes-discovery/pkg/apis/apifederation/types.go
> <https://github.com/kubernetes/kubernetes/pull/37561>:
>
> > +	// InternalHost is the host:port for locating the server inside the pod/service network.
> +	InternalHost string
> +	// Group is the API group name this server hosts
> +	Group string
> +	// Version is the API version this server hosts
> +	Version string
> +
> +	// InsecureSkipTLSVerify disables TLS certificate verification when communicating with this server.
> +	// This is strongly discouraged.  You should use the CABundle instead.
> +	InsecureSkipTLSVerify bool
> +	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
> +	CABundle []byte
> +
> +	// Priority controls the ordering of this API group in the overall discovery document that gets served.
> +	// Client tools like `kubectl` use this ordering to derive preference, so this ordering mechanism is important.
> +	// Secondary ordering is performed based on name.
>
> Also, state the default. I suggest defaulting to the lowest priority/Tier.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37561>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7YaRBg6LBEjPex3_oBogphqtZN5ks5rFKQ3gaJpZM4K-CPK>
> .
>

lavalamp commented:
Fair enough. I care more about the examples etc than the word.

On Mon, Dec 5, 2016 at 4:29 PM, Clayton Coleman <notifications@github.com>
wrote:

> "Tier" feels weird and unusual (to a casual user). Weight might be a good
> generic word. But I'm ok with priority.
>
> On Mon, Dec 5, 2016 at 6:54 PM, Daniel Smith <notifications@github.com>
> wrote:
>
> > *@lavalamp* commented on this pull request.
> > ------------------------------
> >
> > In cmd/kubernetes-discovery/pkg/apis/apifederation/types.go
> > <https://github.com/kubernetes/kubernetes/pull/37561>:
> >
> > > + // InternalHost is the host:port for locating the server inside the
> pod/service network.
> > + InternalHost string
> > + // Group is the API group name this server hosts
> > + Group string
> > + // Version is the API version this server hosts
> > + Version string
> > +
> > + // InsecureSkipTLSVerify disables TLS certificate verification when
> communicating with this server.
> > + // This is strongly discouraged. You should use the CABundle instead.
> > + InsecureSkipTLSVerify bool
> > + // CABundle is a PEM encoded CA bundle which will be used to validate
> an API server's serving certificate.
> > + CABundle []byte
> > +
> > + // Priority controls the ordering of this API group in the overall
> discovery document that gets served.
> > + // Client tools like `kubectl` use this ordering to derive preference,
> so this ordering mechanism is important.
> > + // Secondary ordering is performed based on name.
> >
> > Also, state the default. I suggest defaulting to the lowest
> priority/Tier.
> >
> > —
> > You are receiving this because you are on a team that was mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/37561>, or mute the
> thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_p7YaRBg6LBEjPex3_
> oBogphqtZN5ks5rFKQ3gaJpZM4K-CPK>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37561#issuecomment-265023630>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglq-gOqJFbhf8P1dAEeM7CYZW311lks5rFKyDgaJpZM4K-CPK>
> .
>

deads2k committed:
add apiregistration types

deads2k committed:
generated

deads2k commented:
comments addressed.

deads2k commented:
@k8s-bot unit test this: issue #37961

deads2k commented:
@k8s-bot gci gke e2e test this

deads2k commented:
@k8s-bot cvm gce e2e test this: issue #38198

deads2k commented:
@k8s-bot unit test this

rmmh commented:
@k8s-bot gci gce e2e test this (job was accidentally deleted)
